File: src/rainbow/outer/frame/MainFrame.java
package rainbow.outer.frame;

import rainbow.inner.system.MySystem;
import rainbow.outer.painter.BackPainter;
import rainbow.outer.painter.CoordinateSystemPainter;
import rainbow.outer.painter.MyPainter;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ComponentAdapter;
import java.awt.event.ComponentEvent;
import java.util.ArrayList;

import static rainbow.outer.frame.tool.FrameLocationSetter.center;

/**
 * 主界面
 *
 * @author Rainbow Yang
 */
public class MainFrame extends JFrame {
    private int x;
    private int y;

    public static MainFrame mainFrame;

    public MainFrame() throws HeadlessException {
        mainFrame = this;
        MySystem.getSystem().getVersion().setTitle(this);
        MySystem.getSystem().getSize().setSize(this);
        center(this);
        MySystem.getSystem().getListeners().setListeners(this);

        //todo 将来用View
        add(new AllPainter());


        addComponentListener(new ComponentAdapter() {
            @Override
            public void componentResized(ComponentEvent e) {
                Dimension d = e.getComponent().getSize();

                MySystem.getSystem().getSize().setWidth(d.getWidth());
                MySystem.getSystem().getSize().setHeight(d.getHeight());

                repaint();
            }
        });

        setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);

        setVisible(true);
    }

    class AllPainter extends JPanel {
        private ArrayList<MyPainter> painters = new ArrayList<>();

        //TODO 暂时
        {
            painters.add(new BackPainter());
            painters.add(new CoordinateSystemPainter());
        }

        @Override
        protected void paintComponent(Graphics g) {
            //抗锯齿
            ((Graphics2D) g).setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
            painters.forEach(painters -> painters.paint(g));
        }
    }
}

File: src/rainbow/outer/frame/MainFrame.kt
package rainbow.outer.frame

import rainbow.inner.system.*
import rainbow.inner.system.SystemPainters.repaint
import rainbow.outer.frame.tool.FrameLocationSetter.center
import java.awt.Graphics
import java.awt.event.ComponentAdapter
import java.awt.event.ComponentEvent
import javax.swing.JFrame
import javax.swing.JPanel
import javax.swing.WindowConstants

/**
 * @author Rainbow Yang
 */
class MainFrame : JFrame() {
    companion object {
        var mainFrame: JFrame? = null
    }

    init {
        mainFrame = this

        MySystem.setTitle(this)
        MySystem.setSize(this)
        center(this)

//        SystemPainters.repainter = this::repaint

        SystemListeners.setListeners(this)

        addComponentListener(object : ComponentAdapter() {
            override fun componentResized(e: ComponentEvent) {
                val d = e.component.size
                MySystem.width = d.width.toDouble()
                MySystem.height = d.height.toDouble()
                SystemPainters.repaint()
            }
        })

        add(object : JPanel() {
            override fun paintComponent(g: Graphics) {
                SystemPainters.paint(g)
            }
        })

        defaultCloseOperation = WindowConstants.EXIT_ON_CLOSE
        isVisible = true
    }
}